### PR TITLE
EN-3755, EN-3813 remote configurations, true ip headers, sensitive route bug

### PR DESCRIPTION
**Pull request by nitzanpx**
refactor handle verification
decouple validators from doing http requests by creating pxClient
added uuid to page_requested

### CODE REVIEW

**Comment by yaronpx** [comment]
As usual, indentation is messed up

File: PerimeterXModule/DataContracts/Configuration/PXConfigurationWrapper.cs [hunk 1]
```diff
@@ -0,0 +1,93 @@
...
+using System.Diagnostics;
+
+namespace PerimeterX
+{
+    public class PXConfigurationWrapper
+    {
+        public PXConfigurationWrapper(PxModuleConfigurationSection config)
+        {
+            Enabled = config.Enabled;
+			AppId = config.AppId;
```
Comment (pxjohnny) [inline PerimeterXModule/DataContracts/Configuration/PXConfigurationWrapper.cs]: indentation issue

File: PerimeterXModule/DataContracts/Configuration/PXConfigurationWrapper.cs [hunk 2]
```diff
@@ -0,0 +1,93 @@
...
+			CookieKey = dynamicConfig.CookieKey;
+			BlockingScore = dynamicConfig.BlockingScore;
+			MonitorMode = dynamicConfig.ModuleMode.Equals("monitoring");
+			SocketIpHeader = PxConstants.ArrayToStringCollection(dynamicConfig.IpHeaders);
+			SensitiveHeaders = PxConstants.ArrayToStringCollection(dynamicConfig.SensitiveHeaders);
+			ApiTimeout = dynamicConfig.RiskTimeout;
+			Checksum = dynamicConfig.Checksum;  
+        }
+
+		public bool Enabled { get; set; }
```
Comment (pxjohnny) [inline PerimeterXModule/DataContracts/Configuration/PXConfigurationWrapper.cs]: getter and setters should be on top of the class, will be easier to read.

File: PerimeterXModule/DataContracts/Configuration/PXDynamicConfiguration.cs [hunk 3]
```diff
@@ -0,0 +1,44 @@
...
+		public int BlockingScore { get; set; }
+
+		[DataMember(Name = "appId")]
+		public string AppId { get; set; }
+
+		[DataMember(Name = "moduleMode")]
+		public string ModuleMode { get; set; }
+
+		[DataMember(Name = "ipHeaders")]
+        public string[] IpHeaders { get; set; }
```
Comment (pxjohnny) [inline PerimeterXModule/DataContracts/Configuration/PXDynamicConfiguration.cs]: indendation

File: PerimeterXModule/Internals/CookieDecoder.cs [hunk 4]
```diff
@@ -19,23 +19,25 @@ public string Decode(string cookie)
...
-				throw new NullReferenceException("cookieKey");
+				throw new NullReferenceException("PXConfigurationWrapper");
 			}
-			this.cookieKey = cookieKey;
+            this.pxConfig = pxConfig;
 		}
 
 		public string Decode(string cookie)
 		{
+            var cookieKeyBytes = Encoding.UTF8.GetBytes(pxConfig.CookieKey);
```
Comment (pxjohnny) [inline PerimeterXModule/Internals/CookieDecoder.cs]: Why did you change the cookie key into bytes? if its needed then how did it work before?

File: PerimeterXModule/Internals/PXClient/DefaultPxClient.cs [hunk 5]
```diff
@@ -0,0 +1,94 @@
...
+			// Set request
+			var requestMessage = new HttpRequestMessage(HttpMethod.Get, requestUrl);
+			CancellationTokenSource source = new CancellationTokenSource(TimeSpan.FromMilliseconds(pxConfig.ApiTimeout));
+			CancellationToken cancellationToken = source.Token;
+			var httpResponse = httpClient.SendAsync(requestMessage, cancellationToken).Result;
+
+            Debug.WriteLine(string.Format("response for {0} received status code {1}", requestUrl, httpResponse.StatusCode));
+
+			PXDynamicConfiguration pxDynamicConfiguration = null;
+            if (!httpResponse.StatusCode.Equals(HttpStatusCode.OK))
```
Comment (pxjohnny) [inline PerimeterXModule/Internals/PXClient/DefaultPxClient.cs]: the else if is not needed. if you check for !statuscode.ok then the else can only be statuscode.ok. In addition the return should be inside the if since the only other option is the else (which has return in it)

File: PerimeterXModule/Internals/PxContext.cs [hunk 6]
```diff
@@ -105,21 +105,26 @@ public PxContext(HttpContext context, PxModuleConfigurationSection pxConfigurati
 
 			Ip = context.Request.UserHostAddress;
 			// Get IP from custom header
-			string socketIpHeader = pxConfiguration.SocketIpHeader;
-			if (!string.IsNullOrEmpty(socketIpHeader))
-			{
-				var headerVal = context.Request.Headers[socketIpHeader];
-				if (headerVal != null)
+			var socketIpHeaders = pxConfiguration.SocketIpHeader;
+            if (socketIpHeaders != null){
```
Comment (pxjohnny) [inline PerimeterXModule/Internals/PxContext.cs]: why did you change the isNullOrEmpty to just !=null?

File: PerimeterXModule/Internals/VerificationHandlers/DefaultVerificationHandler.cs [hunk 7]
```diff
@@ -0,0 +1,113 @@
...
+            application.CompleteRequest();
+            return false;
+		}
+
+		private void PostActivity(PxContext pxCtx, PXConfigurationWrapper pxConfig, string eventType, ActivityDetails details = null)
+		{
+			var activity = new Activity
+			{
+				Type = eventType,
+				Timestamp = Math.Round(DateTime.UtcNow.ToUniversalTime().Subtract(new DateTime(1970, 1, 1, 0, 0, 0, DateTimeKind.Utc)).TotalMilliseconds, MidpointRounding.AwayFromZero),
```
Comment (pxjohnny) [inline PerimeterXModule/Internals/VerificationHandlers/DefaultVerificationHandler.cs]: is there no better way of doing the timestamp calculation then using the 1970 hard coded date?

File: PerimeterXModule/Internals/CookieDecoder.cs (hunk 4 above)
Comment (nitzanpx) [inline PerimeterXModule/Internals/CookieDecoder.cs]: Its because before it was receiving the cookie key in bytes on the constructor and it called it cookieKey 
`public EncryptedCookieDecoder(byte[] cookieKey)`
which  was actually this code
`var cookieKeyBytes = Encoding.UTF8.GetBytes(pxConfig.CookieKey);`

File: PerimeterXModule/Internals/VerificationHandlers/DefaultVerificationHandler.cs (hunk 7 above)
Comment (nitzanpx) [inline PerimeterXModule/Internals/VerificationHandlers/DefaultVerificationHandler.cs]: its an old code, is there a better way to set the cookie as expired?
